Creating rust dependencies tree explorer

bruno-ortiz opened the pull request:
Hello! 

I tried to implement a tree view that shows the dependencies of a project.

It allows to see all dependencies to the project and it uses `cargo tree` for it. Also it allows to click and open the files, the viewtree tries its best to follow the openned file in the editor.

Here is an example:
![image](https://user-images.githubusercontent.com/5748995/155822183-1e227c7b-7929-4fc8-8eed-29ccfc5e14fe.png)

Any feedback is welcome since i have basically no professional experience with TS.

bruno-ortiz commented:
Hey guys! I don't want to be a annoying or anything like that, but as this is my first PR here, do i need to do something else for the PR to be reviewed?

Veykril reviewed (commented):
I don't know my way around the client code too much, and I don't think any of the active people do either but overall this seems good to me.

Though I have two questions, how often does this invoke cargo (as in when) and we should ideally not have this code show error popups I think, if this code errors but the rest of r-a works fine I wouldn't personally want to bother users with it. So ideally errors should only be logged, and preferably in a way that is not too spammy.

bjorn3 commented:
Instead of using `cargo tree` should this use the dependency graph of rust-analyzer? That will also make it work when using `rust-project.json` instead of cargo and when using `rust-analyzer.linkedProjects`.

bruno-ortiz commented on editors/code/src/ctx.ts:
```diff
@@ -42,7 +45,24 @@ export class Ctx {
+        const res = new Ctx(config, extCtx, client, serverPath, statusBar, dependenciesProvider, treeView);
+        res.pushCleanup(treeView);
+
+        vscode.window.onDidChangeActiveTextEditor(e => {
+            if (e && isRustEditor(e)) {
+                execRevealDependency(e).catch(reason => {
+                    void vscode.window.showErrorMessage(`Dependency error: ${reason}`);
+                });
+            }
+        });
```
@Veykril Thanks for the review, i'll try to answer your questions:
1- It calls cargo tree is invoked every time the editor opens and after that for every "ok" that it receives from the server: The code is this:
```rust
 switch (status.health) {
            case "ok":
                this.statusBar.color = undefined;
                this.dependencies.refresh(); // it tells the editor that something changed in the server and it should update the dependency tree
                break;
            case "warning":
                this.statusBar.tooltip += "\nClick to reload.";
 ....
```
2- I agree that we should not show the popup errors. I can change that, what are the best practices to logging in the editor?
3- This is executed every time, so the dependency tree can "follow" the user. So every time the user changes the active editor, the dependency tree "follows" the user and reveals the file in it.

bruno-ortiz replied:
@lnicola it does not run cargo tree every time the editor changes, it runs when it receives a "status change" notification from the server. It checks that the status is "ok" and tries to reload the dependencies. 
But as pointed in the other comments ill try to look into the dependency graph feature.

bruno-ortiz commented:
@bjorn3 Can you please send me a reference or any example on how i can use the dependency graph?  And is this blocking, can we make this change afterwards?

Thanks for the comments

bruno-ortiz commented on editors/code/src/commands.ts:
```diff
@@ -246,6 +247,44 @@ export function openCargoToml(ctx: Ctx): Cmd {
+
+export function revealDependency(ctx: Ctx): Cmd {
+    return async (editor: RustEditor) => {
+        const rootPath = vscode.workspace.workspaceFolders![0].uri.fsPath;
```
I'm not too familiar with VSCode API, i'll look into that.

bruno-ortiz commented on editors/code/src/dependencies_provider.ts:
```diff
@@ -0,0 +1,154 @@
+        return [stdlib].concat(deps);
+    }
+
+    private async getStdLib(): Promise<Dependency> {
```
Sure, can you please point me out on where can i find this logic in the server? ty

lnicola replied:
There's https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/view_crate_graph.rs for "View crate graph" and https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/project_model/src/workspace.rs#L140 takes care of loading a workspace. The sysroot is detected by `Sysroot::load` via `Sysroot::discover`.

nemethf replied:
> This duplicates the logic in the server. I think we should do as much as possible there instead of in the client code.

I'm new here, but I disagree.  The server-side implementation takes some of the burden off of the developers of non-VSCode clients.

nemethf replied:
I as well think it should be on the server side.  I'm sorry if I misunderstood something or if I wasn't clear.  Thanks.

Veykril commented:
Ye, the server notifying the client here will have to be done as an lsp extension, a good inspiration for that would probably be the ServerStatus notification, as well as the view dependency graph as noted earlier.

bors commented:
:umbrella: The latest upstream changes (presumably #12294) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Veykril commented:
Went ahead and rebased this, this still needs to use the server for querying, I added a small commit implementing the scaffolding part for the lsp extension for now.

bors commented:
:umbrella: The latest upstream changes (presumably #12919) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bruno-ortiz commented on crates/ide/src/fetch_crates.rs:
```diff
@@ -0,0 +1,59 @@
+        .map(|it| it.canonical_name().to_owned())
+        .unwrap_or("unknown".to_string())
+}
+
```
I feel that the function below is SUPER hacky, but I couldn't figure another way around it.

bruno-ortiz commented:
@Veykril sorry for the long delay, I got swamped at work.

I implemented the server query, feel free to review the code again..

bruno-ortiz commented on crates/ide/src/fetch_crates.rs:
```diff
@@ -0,0 +1,65 @@
+        .unwrap_or("unknown".to_string())
+}
+
+fn crate_path(db: &RootDatabase, data: &ide_db::base_db::CrateData, crate_name: &str) -> String {
+    let source_root_id = db.file_source_root(data.root_file_id);
+    let source_root = db.source_root(source_root_id);
+    let source_root_path = source_root.path_for_file(&data.root_file_id);
+    match source_root_path.cloned() {
+        Some(mut root_path) => {
+            let mut crate_path = "".to_string();
+            while let Some(vfs_path) = root_path.parent() {
+                match vfs_path.name_and_extension() {
+                    Some((name, _)) => {
+                        if name.starts_with(crate_name) {
+                            crate_path = vfs_path.to_string();
+                            break;
+                        }
+                    }
+                    None => break,
+                }
+                root_path = vfs_path;
+            }
+            crate_path
+        }
+        None => "".to_owned(),
+    }
```
I feel that this function is SUPER hacky, but I couldn't figure out another way around it.

bruno-ortiz replied:
@davidbarsky this function tries to find a crate root dir in the filesystem, through the CrateGraph I can get the root file of a crate, for example, tokio's root file is `/home/<user>/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/lib.rs`. But I don't need the root file i need the root dir of a crate, so I need to convert this file to `/home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0`.
If there is a cleaner way of doing it, I'll gladly change the implementation.

nemethf commented on docs/dev/lsp-extensions.md:
```diff
@@ -839,3 +839,26 @@ export interface Diagnostic {
+    crates: {
+        name: string;
+        version: string;
+        path: string;
```
Wouldn't it be more general (and more in line with the language server protocol) to use `textDocument: TextDocumentIdentifier` here instead of the current `path: string`?

And as a side note: why this is called a graph/tree, when this is just a list?  It is not possible to construct a graph from the response.  Ie.: visualize the the dependencies of a dependency.

bruno-ortiz replied:
@nemethf you're right about the naming, I changed from "graph" to "list". The tree is because the crates are displayed in a "ViewTree".
About the `TextDocumentIdentifier `, I tried using it, but it just made the code more complicated, because to reveal the items in the viewTree I need to work with the filesystem path, and the `TextDocumentIdentifier` inputs the "scheme" in the path.

nemethf replied:
> The tree is because the crates are displayed in a "ViewTree". 

Thank you.  I have no objections against it.  However, it is a bit strange that a client-side implementation detail determines an abstract name in protocol specification.  I can imagine some clients using this extension for something else, for example, to provide a "jump to dependency" functionality that implements an incremental search on the dependencies without showing them beforehand. 

Also this fetchDependencyGraph reminds me of the rust-analyzer/viewCrateGraph LSP extension.  In the long run, it might make sense to generalize both of them to send the same details.   fetchDependencyGraph could list the dependencies of a dependency, and viewCrateGraph could be extended to have a uri/path for each dependency.

Veykril replied:
A `TextDocumentIdentifier` would indeed make more sense if possible

davidbarsky commented on editors/code/src/commands.ts:
```diff
@@ -260,6 +261,33 @@ export function openCargoToml(ctx: CtxInit): Cmd {
+    return async (editor: RustEditor) => {
+        const rootPath = vscode.workspace.workspaceFolders![0].uri.fsPath;
+        const documentPath = editor.document.uri.fsPath;
+        if (documentPath.startsWith(rootPath)) return;
```
I'm not entirely sure why you'd want to do this. There's already non-workspace filtering done on the server; this just adds an additional check.

(This also breaks for monorepos where users open their editor at the monorepo root. Non-workspace crates are often stored in `third-party/` but share the workspace root.)

bruno-ortiz replied:
The purpose of the function revealDependency is to make the TreeView "reveal" the opened file in the editor. Example:
![image](https://user-images.githubusercontent.com/5748995/229884702-cc765e2a-3e48-440e-9937-b925f2844bd7.png)

But if I where to open a file that belongs to the local project, I dont want to trigger any "revealing"

davidbarsky replied:
> The purpose of the function revealDependency is to make the TreeView "reveal" the opened file in the editor. 

Gotcha, thanks for explaining—I've built the branch locally and usage made a bit more sense.

> But if I where to open a file that belongs to the local project, I dont want to trigger any "revealing"

That makes sense and works for most default Cargo-based builds. However, using `if (documentPath.startsWith(rootPath)) return;` in situations where `cargo vendor` has been run _or_ when a user opens their VS Code instance at the root of the monorepo (where a `third-party/` folder exists and contains vendored Rust crates), this check will prevent this tree-based explorer from ever opening and it seems unfortunate that those users won't benefit from this feature[^1]. I don't have merge permissions on rust-analyzer, but I'd recommend removing this check, as `fetch_crates` in `crates/ide/src/fetch_crates.rs` already performs a similar check. I need to look over this PR a bit more, but if a user asks to reveal the dependency for a crate that is part of their editable workspace, then I would consider showing it. Alternatively, it might make sense to add another LSP extension to query the server itself to check whether a file is editable/part of the current editable workspace or not. 

[^1]: Personally, I'd love to build atop of this UI in order to allow users to incrementally add crates to their (editable) rust-analyzer workspace—it's a pretty workflow in monorepos.

bruno-ortiz replied:
@davidbarsky thanks for the reply, I just pushed a commit improving some pain points and hopefully giving it proper support to multiple workspace folders.
Could you point me at any github repos that make use of the features that you mentioned? I would love to test this cases because I suspect that they won't be a problem, because if they're already part of the project tree, the main treeView of vscode will already display those dependencies.

bruno-ortiz replied:
Just to add to my previous comment, I created a vendored project and everything seems fine.
![image](https://user-images.githubusercontent.com/5748995/230171372-3645d20c-22d8-4c5d-83be-4bb691e2fa40.png)

davidbarsky replied:
> Could you point me at any github repos that make use of the features that you mentioned?

@bruno-ortiz I can't think of any project off-hand, but it would be possible to take, e.g., the rust-analyzer repository and run `cargo vendor` and you'll see a subset of the behavior I'm mentioning.

> I would love to test this cases because I suspect that they won't be a problem, because if they're already part of the project tree, the main treeView of vscode will already display those dependencies.

At the risk of being *that person*, but a smaller repository won't really demonstrate this issue—it only really becomes apparent once a monorepo has tens (or hundreds!) of millions of lines of code. While said dependencies _are_ technically part of the primary explorer view, it's often infeasible to see them at a glance because of the sheer number of folders present in the default file explorer.

My mental model of this feature—at least in monorepos—is that it is a subset of the monorepo (think of it as a materialized view...) that corresponds to the portions of monorepo most relevant to the code being worked on at the moment.

The reason that I'm so adamant about requesting this is that I didn't really _get_ the need/desire for it until I tried it. While it is technically possible to navigate through the primary explorer window to dependencies, it ends up taking closer to 30-60 seconds per dependency rather than something instantaneous.

davidbarsky commented on editors/code/src/commands.ts:
```diff
@@ -260,6 +261,33 @@ export function openCargoToml(ctx: CtxInit): Cmd {
+        } else {
+            let documentPath = editor.document.uri.fsPath;
+            const parentChain: DependencyId[] = [{ id: documentPath.toLowerCase() }];
+            do {
+                documentPath = path.dirname(documentPath);
+                parentChain.push({ id: documentPath.toLowerCase() });
+            } while (!ctx.dependencies?.contains(documentPath));
+            parentChain.reverse();
+            for (const idx in parentChain) {
+                await ctx.treeView?.reveal(parentChain[idx], { select: true, expand: true });
```
do you mind adding a comment explaining why this bit is necessary? is it to show a dependency of a dependency?

bruno-ortiz replied:
I'll add it.
Its because the TreeView is Lazy, so at first it only has the root dependencies: Example:
- core
- alloc
- std

if I want to reveal alloc/src/str.rs, I have to:
1. reveal every children of alloc
- core
- alloc
&emsp;|-beches
&emsp;|-src
&emsp;|- ...
- std
2. reveal every children of src:
- core
- alloc
&emsp;|-beches
&emsp;|-src
&emsp;&emsp;| - lib.rs
&emsp;&emsp;| - str.rs    <------- FOUND IT!
&emsp;&emsp;| - ...
&emsp;|- ...
- std

So basically, i have to calculate the parents of the file that I want to reveal until i find the root crate path.

davidbarsky commented on crates/ide/src/fetch_crates.rs:
```diff
@@ -0,0 +1,71 @@
+        while let Some(vfs_path) = root_path.parent() {
+            match vfs_path.name_and_extension() {
+                Some((name, _)) => {
+                    if name.starts_with(crate_name) {
```
the file path component here isn't necessarily guaranteed to start with crate name. i mentioned this in another point, but I think getting the full URI of the crate would probably make some other parts in this PR easier to implement.

bruno-ortiz replied:
I agree, the problem here is that i could no find the full URI of the crate anywhere, maybe someone with more experience in the RA API could help me.

bruno-ortiz commented:
I converted this PR to draft because I detected some problems working with multi-workspaces. 
I'll try to fix it ASAP.

bruno-ortiz commented:
Marking this as ready to review again, I thought it would take longer to fix the multi-workspace issues.

bruno-ortiz commented:
@davidbarsky @nemethf I've made some quality of life improvements and made some best effort attempts to support the every usecase possible. One pain point is:

I could not figure out if a crate is vendored or not. Since RA uses "cargo metadata" to fetch the info about the crates, it receives this info from cargo itself, so unless we are able to parse cargo's config(config.toml), I think we will not be able to get a hold of this information. This doesn't mean that it doesn't works for projects with vendored crates, the dependencies will be shown, but the automatic navigation will be turned off, since vscode itself will auto-navigate using the primary explorer.

Does anyone see any other issues that would make this PR unmergeable?

nemethf commented:
> Does anyone see any other issues that would make this PR unmergeable?

I've only looked at lsp-extensions.md, and it is not a showstopper for me, but there is a bit of an inconsistency: the fetchDependencyList has FetchDependencyGraphParams and FetchDependencyGraphResult.  I think it should be FetchDependencyListParams and FetchDependencyListResult.

Thank you.

Veykril commented on crates/base-db/src/input.rs:
```diff
@@ -304,6 +304,7 @@ pub struct CrateData {
     /// For purposes of analysis, crates are anonymous (only names in
     /// `Dependency` matters), this name should only be used for UI.
     pub display_name: Option<CrateDisplayName>,
+    pub crate_root_path: Option<AbsPathBuf>,
```
We shouldn't store any paths like this in `CrateData`, instead we should resolve the `root_file_id` back to its path where we need it

bruno-ortiz replied:
Ok, I was unsure that this was the right aproach, but could you elaborate a little more on why this is bad?

Veykril replied:
We want as little code to know about paths as possible. Paths are tricky to handle correctly as different paths can describe the same file etc, FileId's are unambiguous. And in this case, we already store the file id, so storing the path as well is redundant.

Veykril commented on crates/ide/src/fetch_crates.rs:
```diff
@@ -0,0 +1,45 @@
+pub struct CrateInfo {
+    pub name: String,
+    pub version: String,
+    pub path: String,
```
This should probably be a `FileId` instead then (given the previous comment),

Veykril commented on crates/paths/src/lib.rs:
```diff
@@ -213,6 +213,13 @@ impl AbsPath {
         self.0.exists()
     }
+
+    pub fn name_and_extension(&self) -> Option<(&str, Option<&str>)> {
+        Some((
+            self.file_stem()?.to_str()?,
+            self.extension().and_then(|extension| extension.to_str()),
+        ))
+    }
```
This isn't a delegated method so we shouldn't put that into the `region` comment area, though assuming this is only in one place we might as well replace the function call with what we do here

Veykril commented on crates/rust-analyzer/src/handlers.rs:
```diff
@@ -125,6 +128,19 @@ pub(crate) fn handle_memory_usage(state: &mut GlobalState, _: ()) -> Result<Stri
+    Ok(FetchDependencyListResult {
+        crates: crates
+            .into_iter()
+            .map(|it| CrateInfoResult { name: it.name, version: it.version, path: it.path })
```
Here we need to resolve the file id to its path, or url even, with the `GlobalStateSnapshot`

Veykril commented on crates/rust-analyzer/tests/slow-tests/tidy.rs:
```diff
@@ -45,7 +45,7 @@ fn check_lsp_extensions_docs() {
         let text = lsp_extensions_md
             .lines()
-            .find_map(|line| line.strip_prefix("lsp_ext.rs hash:"))
+            .find_map(|line| dbg!(line.strip_prefix("lsp_ext.rs hash:")))
```
leftover dbg

Veykril commented on editors/code/src/dependencies_provider.ts:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    private async getRootDependencies(): Promise<Dependency[]> {
+        const dependenciesResult: FetchDependencyListResult = await this.ctx.client.sendRequest(
```
when exactly is this being fired? This should ideally only trigger when something relevant to project changes

Veykril commented:
I believe this will show duplicate entries for project that have multiple cargo workspaces/linked projects loaded. Not a huge problem but something we should be aware of (this would be fixed once we deduplicate crates in the crate graph).

Veykril commented on crates/rust-analyzer/src/handlers.rs:
```diff
@@ -1916,3 +1939,24 @@ fn run_rustfmt(
     }
 }
+
+//Thats a best effort to try and find the crate path
+fn crate_path(crate_name: Option<&String>, root_file_path: VfsPath) -> Option<VfsPath> {
+    crate_name.and_then(|crate_name| {
+        let mut crate_path = None;
+        let mut root_path = root_file_path;
+        while let Some(path) = root_path.parent() {
+            match path.name_and_extension() {
+                Some((name, _)) => {
+                    if name.starts_with(crate_name.as_str()) {
+                        crate_path = Some(path);
+                        break;
+                    }
+                }
+                None => break,
+            }
+            root_path = path;
+        }
+        crate_path
+    })
+}
```
This at least deserves a comment explaining what it does as it is not clear at all. Though we could be smarter about this I think. 
We can iterate over all cargo workspaces we know, iterate over the pacakages there, then iterate over the targets that belong to those packages and check if the targets root file path is the same as the path we are looking for. We now have a package with a fitting root path, from there we can read out the cargo toml path of the package and go up to the dir.

(this is obviously quadratic so we probably want to do some upfront hashmap collecting once to reduce the iteration amount required.

Veykril commented on crates/rust-analyzer/src/lsp_ext.rs:
```diff
@@ -27,6 +27,31 @@ pub struct AnalyzerStatusParams {
+pub struct CrateInfoResult {
+    pub name: Option<String>,
+    pub version: Option<String>,
+    pub path: String,
```
was changing this a to `TextDocumentIdentifier` still not possible?

bruno-ortiz replied:
I really don't think that this should be a TextDocumentIdentifier, because it represents the path to the crate's directory, not to a text document.

Veykril commented:
Also please remove the merge commits, we prefer to not introduce merge commits aside from the ones bors introduces.

bruno-ortiz commented:
@Veykril I added the requested comment and rebased the commits onto master branch.

Veykril commented on crates/rust-analyzer/src/handlers.rs:
```diff
@@ -1910,3 +1932,34 @@ fn run_rustfmt(
         Ok(Some(to_proto::text_edit_vec(&line_index, diff(&file, &new_text))))
     }
 }
+
+/// Searches for the directory of a Rust crate with a given name in the directory tree
+/// of the root file of that crate.
+///
+/// # Arguments
+///
+/// * `crate_name`: The name of the crate to search for. This should be a `Some` value if
+///   a crate name has been specified, or `None` if no crate name has been specified.
+/// * `root_file_path`: The path to the root file of the crate.
+///
+/// # Returns
+///
+/// An `Option` value representing the path to the directory of the crate with the given
+/// name, if such a crate is found. If no crate with the given name is found, this function
+/// returns `None`.
+fn crate_path(crate_name: Option<&String>, root_file_path: VfsPath) -> Option<VfsPath> {
+    crate_name.and_then(|crate_name| {
+        let mut root_path = root_file_path;
+        while let Some(path) = root_path.parent() {
+            if let Some((name, _)) = path.name_and_extension() {
+                if name.starts_with(crate_name.as_str()) {
+                    return Some(path);
+                }
+            } else {
+                break;
+            }
+            root_path = path;
+        }
+        None
+    })
+}
```
Okay, so this will not work too well with crates that are renamed in the cargo toml causing their folder name to differ from the crate name. I think a better approach here is to assume the classic library structure, go up two steps and check whether the `Cargo.toml` is where we expect it. That is check that `{parentPath}/../Cargo.toml` exists, then return `{parentPath}/../..`, so we assume `../crate/src/lib.rs` as the crate root (which is almost always the case for libraries)

bruno-ortiz replied:
nice, I changed the code meet your suggestion, it made the code simpler 👍

Veykril commented:
Two more things then I think its ready to go 🎉 
(though I'll wait with merging until next week just to have some nightly time for it)

bruno-ortiz commented:
@Veykril I made the requested fixes:

* changed the type of the returned path to be an Url;
* changed the crate_path function to lookup for the Cargo.toml dir.

bors commented:
:umbrella: The latest upstream changes (presumably #14664) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:v: @bruno-ortiz can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"Veykril","delegate":"bruno-ortiz"} -->

bruno-ortiz commented:
@Veykril I fixed the merging conflicts but I don't have merge permissions.

bruno-ortiz commented:
@bors r+

bors commented:
:hourglass: Testing commit ecfe7c04888a9c2567773370b127d2e6e1cdaa22 with merge a48e0e14e15abf47feae17e54d149eb443729375...
<!-- homu: {"type":"BuildStarted","head_sha":"ecfe7c04888a9c2567773370b127d2e6e1cdaa22","merge_sha":"a48e0e14e15abf47feae17e54d149eb443729375"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/4862590640/jobs/8669460422)
Approved by: bruno-ortiz
Pushing a48e0e14e15abf47feae17e54d149eb443729375 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bruno-ortiz","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/4862590640/jobs/8669460422"},"merge_sha":"a48e0e14e15abf47feae17e54d149eb443729375"} -->
